### DISCUSSION TITLE
Question

### DISCUSSION THREAD

**Discussion by hm-ch** [elsa-workflows/elsa-core#1400] [Q&A]
Hi  @sfmskywalker 

First of all i would like to thank you about the great job.
Afterr publishing the workflow using workflow designer, how can I start a new workflow instance based on WorkflowDefinitionId or Workflow Name.

Thanks

**Comment by sfmskywalker** [comment]
Thanks!

To start a workflow, you have two options:

1. Via the REST API
2. Using C# code (that you executed from within the workflow server application).

### API

To use the REST API, you need to have a workflow definition ID. When you have it (e.g. `7b626d125eae49ce99e30b4f223322aa`), you can execute the workflow like this:

```
curl --location --request POST 'https://localhost:11000/v1/workflows/7b626d125eae49ce99e30b4f223322aa/execute' \
--header 'Content-Type: application/json' \
--data-raw '{
}'
```

That will execute the workflow synchronously. If you want to execute the workflow from a background job instead (so that the HTTP call returns even before the workflow starts, you can **dispatch** the workflow instead, like this:

```
curl --location --request POST 'https://localhost:11000/v1/workflows/7b626d125eae49ce99e30b4f223322aa/dispatch' \
--header 'Content-Type: application/json' \
--data-raw '{
}'
```

For either call, you can provide input to the workflow by including a JSON payload with the `"input"` field.
For example:

```
curl --location --request POST 'https://localhost:11000/v1/workflows/7b626d125eae49ce99e30b4f223322aa/dispatch' \
--header 'Content-Type: application/json' \
--data-raw '{
    "input": "Demo Input"
}'
```

### C#

If you want to execute a workflow programmatically from within the workflow server, then you use the `IWorkflowLaunchpad` that the REST APIs use internally.

For example, to execute a workflow synchronously:

```csharp
var workflowDefinitionId = "7b626d125eae49ce99e30b4f223322aa";
await _workflowLaunchpad.FindAndExecuteStartableWorkflowAsync(workflowDefinitionId);
```

To dispatch the workflow, you need to perform two steps: find the workflow to dispatch, then dispatch:

```csharp
var startableWorkflow = await _workflowLaunchpad.FindStartableWorkflowAsync(workflowDefinitionId);
var result = await _workflowLaunchpad.DispatchStartableWorkflowAsync(startableWorkflow);
```

**Reply by Nokecy** [reply]
Use c# , What is the difference between synchronously and dispatch @sfmskywalker

**Reply by sfmskywalker** [reply]
**Synchronously**

When executing a workflow synchronously, it means that the when the call to e.g. `await _workflowLaunchpad.FindAndExecuteStartableWorkflowAsync(workflowDefinitionId)` completes. the workflow will have executed (until the first blocking activity or until the workflow reached the end).

**Dispatch**

When dispatching a workflow for execution, it means that the call to e.g. `var result = await _workflowLaunchpad.DispatchStartableWorkflowAsync(startableWorkflow)` completes before the workflow has even started.

What happens inside this call is that it will schedule a message onto a queue and then returns control to the caller.
Meanwhile, the message will (eventually) be picked up from a background job (a queue worker, aka consumer) that will then execute the workflow.
